### PR TITLE
osd: optimize object cache and cache pool

### PR DESCRIPTION
**Pull request by XinzeChi**
such as in agent_work, recovery

Signed-off-by: Xinze Chi <xmdxcxz@gmail.com>

### CODE REVIEW

File: src/osd/ReplicatedPG.h [hunk 1]
```diff
@@ -970,11 +971,12 @@ class ReplicatedPG : public PG, public PGBackend::Listener {
...
 protected:
 
-  ObjectContextRef create_object_context(const object_info_t& oi, SnapSetContext *ssc);
+  ObjectContextRef create_object_context(const object_info_t& oi, SnapSetContext *ssc, bool nocache = false);
   ObjectContextRef get_object_context(
     const hobject_t& soid,
     bool can_create,
-    map<string, bufferlist> *attrs = 0
+    map<string, bufferlist> *attrs = 0,
+    bool nocache = false
```
Comment (liewegas) [inline src/osd/ReplicatedPG.h]: I would be inclined to put nocache after can_create, and have no default value, so the cache/nocache is explicit in every caller.  And make true -> cache, false -> don't cache, as @tchaikov suggested!arg is e

**Comment by liewegas** [comment]
If I'm reading the test correct the first key = 1 validates that we get back the stored value even with nocache as long as any reference is alive.  Correct?   Just want to be certain as that is the OSD's requirement!

**Comment by XinzeChi** [comment]
@liewegas 
cache.lookup_or_create(key, true); // nocache the key
cache.lookup(key) // this would cache the key, so if we want to nocache this key, we should use cache.lookup(key, true) which we have added the test case which key = 3

**Comment by XinzeChi** [comment]
ping @liewegas. please help to review.
Frankly speaking,  i am not sure whether or not we should cache obc in some case.
Thanks for any hint.

**Comment by athanatos** [comment]
This needs to be rebased.  I'm also not at all sure this will help much.  The main purpose of the obc cache is merely to capture the really obvious cases.  I'm not sure I want to deal with the complexity of making it smarter.  I also really do not want to rely on the object context cache to infer recency.

**Pull request closed without merging**